### PR TITLE
[BEAM-570] Several enhancements and deprecations to Beam regarding Compression (part 2 of 2)

### PR DESCRIPTION
**Pull request by katsiapis**
- Getting rid of CompressionTypes.ZLIB and CompressionTypes.NO_COMPRESSION.
- Introducing BZIP2 compression in analogy to Dataflow Java's BZIP2, towards resolution of https://issues.apache.org/jira/browse/BEAM-570.
- Introducing SNAPPY codec support for AVRO conciseness and in order to fully resolve https://issues.apache.org/jira/browse/BEAM-570.
- Moving avroio from compression_type to codec as per various discussions.
- A few cleanups in avroio.
- Making textio more DRY and doing a few cleanups.
- Raising exceptions when splitting is requested for compressed source since that should never happen (guaranteed by the service for the supported compression types).
- Using cStringIO instead of StringIO in various places as decided in some other discussions.

### CODE REVIEW

**Comment by katsiapis** [comment]
R: @chamikaramj, can you please take a look?

**Comment by katsiapis** [comment]
R: @robertwb can you please take a look?

I've merged all the changes with @tim1357's PR which should be merged into (https://github.com/apache/incubator-beam/pull/1007)

**Comment by chamikaramj** [comment]
LGTM

**Comment by katsiapis** [comment]
Thanks @chamikaramj!
@robertwb, could you merge this into Beam?

File: sdks/python/apache_beam/io/fileio.py [hunk 1]
```diff
@@ -249,14 +230,11 @@ def __exit__(self, exception_type, exception_value, traceback):
...
-      # prevented. Here we prevent initial splitting by ignoring all splits
-      # other than the split that starts at byte 0.
-      #
-      # TODO: Turns this warning into an exception soon.
-      logging.warning('Ignoring split starting at (%s) for compressed source.',
-                      self.current_offset)
-      return
+      # When compression is enabled both initial and dynamic splitting should
+      # not be allowed.
+      raise ValueError(
```
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio.py]: Are we sure it never accidentally happens?

File: sdks/python/apache_beam/io/fileio.py [hunk 2]
```diff
@@ -95,18 +87,15 @@ def is_valid_compression_type(cls, compression_type):
...
         cls.BZIP2: 'application/x-bz2',
+        cls.GZIP: 'application/x-gzip',
     }
     return mime_types_by_compression_type.get(compression_type, default)
 
   @classmethod
   def detect_compression_type(cls, file_path):
     """Returns the compression type of a file (based on its suffix)"""
-    compression_types_by_suffix = {'.gz': cls.GZIP,
-                                   '.z': cls.ZLIB,
```
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio.py]: Is the removal of zlib intentional? I don't know how many users there are, but it's not backwards compatible.

File: sdks/python/apache_beam/io/fileio.py [hunk 3]
```diff
@@ -463,13 +437,9 @@ def open(path,
            compression_type=CompressionTypes.AUTO):
     if compression_type == CompressionTypes.AUTO:
       compression_type = CompressionTypes.detect_compression_type(path)
-    elif compression_type == 'AUTO':
-      # TODO: Remove this backwards-compatibility soon.
```
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio.py]: How have we determined that it's now safe to act on these TODOs? We should at least highlight these changes in release notes somewhere, right?

File: sdks/python/apache_beam/io/fileio.py [hunk 4]
```diff
@@ -652,12 +624,23 @@ def _fetch_to_internal_buffer(self, num_bytes):
...
       else:
+        # EOF reached.
+        # Verify completeness and no corruption and flush (if needed by
+        # the underlying algorithm).
+        if self._compression_type == CompressionTypes.BZIP2:
+          # Having unused_data past end of stream would imply file corruption.
+          assert not self._decompressor.unused_data, 'Possible file corruption.'
+          try:
+            # EOF implies that the underlying BZIP2 stream must also have
+            # reached EFO. We expect this to raise an EOFError and we catch it
```
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio.py]: s/EFO/EOF/

File: sdks/python/apache_beam/io/fileio.py [hunk 5]
```diff
@@ -697,8 +680,7 @@ def close(self):
     if self._file is None:
       return
 
-    if self._writeable():
-      self._file.write(self._compressor.flush())
+    self.flush()
```
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio.py]: Why this change?

File: sdks/python/apache_beam/io/fileio_test.py [hunk 6]
```diff
@@ -188,7 +188,7 @@ def test_read_entire_file_bzip2_empty(self):
     self.assertEqual(read_lines, [])
 
   def test_read_entire_file_bzip2_large(self):
-    lines = ['Line %d' % d for d in range(10 * 1000)]
+    lines = ['Line %d' % d for d in range(100 * 1000)]
```
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio_test.py]: How expensive does this make these tests?

File: sdks/python/apache_beam/io/fileio.py (hunk 2 above)
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio.py]: Yes, it is intentional.

We are only supporting the GZIP standard (which is more or less the same compression as ZLIB, just different headers/checksup) but also more widely compatible with external tools.

This also makes Dataflow Java and Dataflow Python in-sync with each other as well as with the Dataflow service which only recognized BZIP2 and GZIP as compressed text files at the moment.

File: sdks/python/apache_beam/io/fileio.py (hunk 5 above)
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio.py]: It removes code duplication.

File: sdks/python/apache_beam/io/fileio.py (hunk 1 above)
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio.py]: Based on old and recent discussions with Chamikara, this would only happen if:
a) The Service introduces initial splits for compressed sources (which it shouldn't do, or else it's a bug).
or
b) The worker dynamically splits a compressed source (which it shouldn't do, or else it's a bug).

The various unit and integration tests (both internal and external) passed, which suggests that both the service and the worker are behaving correctly.

I'm open to suggestions of extra testing as well.

File: sdks/python/apache_beam/io/fileio.py (hunk 4 above)
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio.py]: Done.

File: sdks/python/apache_beam/io/fileio.py (hunk 5 above)
Comment (robertwb) [inline sdks/python/apache_beam/io/fileio.py]: But this means we're calling self._file.flush() when closing non-writable files which seems odd.

File: sdks/python/apache_beam/io/fileio_test.py (hunk 6 above)
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio_test.py]: The sum of all additional tests and changes in this file after this PR seems to add ~10 seconds.
(From 15 seconds before to 25 seconds after).

**Comment by katsiapis** [comment]
Thanks Robert!
I've updated the commit message to be a bit more descriptive (but I understand it should also be short).
My first comment has a more detailed description of the itemized changes. 
Let me know if you had something different in mind.

File: sdks/python/apache_beam/io/fileio.py (hunk 5 above)
Comment (katsiapis) [inline sdks/python/apache_beam/io/fileio.py]: Reverted back based on your other comment regarding read-only files.

**Comment by katsiapis** [comment]
Hey @robertwb, test failures seem unrelated to my CL (but perhaps I misunderstood)? Can you take a look and verify?

**Comment by robertwb** [comment]
Yes, the test failures are due to not having https://github.com/apache/incubator-beam/pull/1110 (which, incidentally, is why I didn't have time to get to this yesterday). 

Generally, it's preferred if the commit messages themselves have the descriptive message (a one-liner title, plus any amount of stuff in the body), with truly separate changes being in separate commits (if not PRs, though that's not required). 

I'll rebase this and make sure all tests pass manually (they did before, so I'm relatively confident they will).

**Comment by katsiapis** [comment]
Thanks Robert, will try to following the separate (descriptive) commits per logical change in future contributions.

**Comment by katsiapis** [comment]
Figured out how to update the commit messages.
The first commit message now has the same details as my first comment in this CL (ie description of the full changes).

**Comment by robertwb** [comment]
Thanks. 

I edited the commit message a bit more (as "Several enhancements to Dataflow" applies to essentially every commit in this repo, and "part 2 of 2" doesn't make much sense without part 1 (and for future reference, the grouping of commits in a PR often makes numbering the parts unneeded).

**Comment by katsiapis** [comment]
Thanks Robert, closing this PR.
